impl/authorization/authorization.go

```diff
@@ -38,30 +42,37 @@ func New() authorization.Authorization {
 // authorized to carry the given permission. A call is authorized if:
 //  1. userID matches the identity in sctx,
 //  2. or, sctx's identity is authorized to do the action in domainID and appID.
-func (a *authz) IsAuthorized(sctx *authentication.SecurityContext,
+func (a *authz) Authorize(ctx context.Context,
 	domainID, appID, userID string, permission authzpb.Permission) error {
+	sctx, ok := authentication.FromContext(ctx)
+	if !ok {
+		return status.Errorf(codes.Unauthenticated, "Request does not contain a ValidatedSecurity object")
+	}
+
 	// Case 1.
-	if sctx.Identity() == userID {
+	if sctx.Email == userID {
 		return nil
 	}
 
 	// Case 2.
 	rLabel := resourceLabel(domainID, appID)
 	roles, ok := a.policy.GetResourceToRoleLabels()[rLabel]
 	if !ok {
-		return fmt.Errorf("resource <domainID=%v, appID=%v> does not have a defined policy", domainID, appID)
+		return status.Errorf(codes.PermissionDenied, "%v does not have a defined policy", rLabel)
 	}
 	for _, l := range roles.GetLabels() {
 		role := a.policy.GetRoles()[l]
-		if isPrincipalInRole(role, sctx.Identity()) && isPermisionInRole(role, permission) {
+		if isPrincipalInRole(role, sctx.Email) && isPermisionInRole(role, permission) {
 			return nil
 		}
 	}
-	return fmt.Errorf("%v is not authorized to perform %v on resource defined by <domainID=%v, appID=%v>", sctx.Identity(), permission, domainID, appID)
+	return status.Errorf(codes.PermissionDenied, "%v is not authorized to perform %v on %v", sctx.Email, permission, rLabel)
 }
 
 func resourceLabel(domainID, appID string) string {
-	return fmt.Sprintf("%v|%v", domainID, appID)
+	return fmt.Sprintf("domains/%v/apps/%v",
+		strings.Replace(domainID, "/", "_", -1),
```

Comment:
this implies that two resources with differing domains (and/or apps) "blah/blah" and "blah_blah" are treated identically, is that ok?

Reply:
The `/` character is not allowed in domainIDs or appIDs.  If they were allowed, a specially crafted domainID or appID could cross boundaries. eg `domainID = "blah/apps/apptakeover`

Reply:
Sure, but my point is that there are now multiple "preimages" which result in identical resource labels - this function doesn't enforce the rule that "/"s aren't allowed, it just quietly papers over that situation resulting in the multiple "preimage" thing.
Maybe that's OK, I don't know :)

Is the rule about the contents of domainID and appID enforced elsewhere? 
if so this function possibly doesn't need to `Sprintf`, and if it isn't enforced elsewhere then maybe this function should return an error in the case where a passed-in ID contains the character?

Reply:
Excellent point. I'll make this return an error.